File: editors/vscode/src/lsp/requests.ts
import { RequestType } from 'vscode-languageclient';

import * as model from './model'


export namespace projects {
    export namespace create {
        export interface Parameters {
            // Path to a directory where project files should be created
            directoryUri: string
            projectName: string
        }
    
        export interface Response {
            // Path to the newly created manifest that should be opened by the client
            manifestUri: string
        }
    
        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/projects/create");
    }

    export namespace list {
        export interface Parameters {
            // Defaults to true
            onlyFromWorkspace?: boolean
        }
    
        export interface Response {
            projectInfos: model.ContentInfo[]
        }
    
        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/projects/list");
    }

    export namespace vanillaDependencyContent {
        export interface Parameters {
            projectUri: string
        }

        export interface Response {
            content0Info: model.ContentInfo
        }

        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/projects/vanillaDependencyContent");
    }

    export namespace vanillaContent {
        export interface Parameters {
            
        }

        export interface Response {
            content0Info: model.ContentInfo | undefined
        }

        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/projects/vanillaContent");
    }

    export namespace sourceTree {
        export interface Parameters {
            contentUri: string
        }

        export interface Response {
            scriptsRootPath: string,
            localScriptPaths: string[]
        }

        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/projects/sourceTree");
    }
}

export namespace scripts {
    export namespace parent_content {
        export interface Parameters {
            scriptUri: string
        }

        export interface Response {
            parentContentInfo: model.ContentInfo
        }

        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/scripts/parentContent");
    }
}

export namespace debug {
    export namespace scriptAst {
        export interface Parameters {
            scriptUri: string
        }
    
        export interface Response {
            ast: string
        }
    
        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/debug/scriptAst");
    }

    export namespace scriptCst {
        export interface Parameters {
            scriptUri: string
        }
    
        export interface Response {
            cst: string
        }
    
        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/debug/scriptCst");
    }

    export namespace contentGraphDot {
        export interface Parameters {

        }

        export interface Response {
            dotGraph: string
        }

        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/debug/contentGraphDot");
    }

    export namespace scriptSymbols {
        export interface Parameters {
            scriptUri: string
        }
    
        export interface Response {
            symbols: string
        }
    
        export const type = new RequestType<Parameters, Response, void>("witcherscript-ide/debug/scriptSymbols");
    }
}

File: crates/lsp/src/requests.rs
use serde::Deserialize;
use serde::Serialize;
use tower_lsp::lsp_types as lsp;
use super::model;


pub mod projects {
    use super::*;

    /// Creates a script project in the specified directory.
    /// Returns path to the newly created manifest file.
    pub mod create {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub directory_uri: lsp::Url,
            pub project_name: String
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub manifest_uri: lsp::Url
        }

        pub const METHOD: &'static str = "witcherscript-ide/projects/create";
    }

    /// Lists information about currently managed projects 
    pub mod list {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            /// If is None, defaults to true
            pub only_from_workspace: Option<bool>
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub project_infos: Vec<model::ContentInfo>
        }

        pub const METHOD: &'static str = "witcherscript-ide/projects/list";
    }

    /// Returns information about content0 dependency of a given project, if it depends on it
    pub mod vanilla_dependency_content {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub project_uri: lsp::Url
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub content0_info: model::ContentInfo
        }

        pub const METHOD: &'static str = "witcherscript-ide/projects/vanillaDependencyContent";
    }

    /// Returns information about content0 content, doesn't need any source content to check as opposed to [`vanilla_dependency_content`].
    /// Returns None if doesn't find content0
    pub mod vanilla_content {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            
        }

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub content0_info: Option<model::ContentInfo>
        }

        pub const METHOD: &'static str = "witcherscript-ide/projects/vanillaContent";
    }

    /// Returns a list of source file paths associated with a given content sorted in alphabetical order
    pub mod source_tree {
        use std::path::PathBuf;
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub content_uri: lsp::Url
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub scripts_root_path: PathBuf,
            pub local_script_paths: Vec<PathBuf>
        }

        pub const METHOD: &'static str = "witcherscript-ide/projects/sourceTree";
    }
}

pub mod scripts {
    use super::*;

    /// Returns information about content containing the given script
    pub mod parent_content {
        use super::*;
    
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub script_uri: lsp::Url
        }
    
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub parent_content_info: model::ContentInfo
        }
    
        pub const METHOD: &'static str = "witcherscript-ide/scripts/parentContent";
    }    
}

pub mod debug {
    use super::*;

    /// Returns script file's AST representation
    pub mod script_ast {
        use super::*;
    
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub script_uri: lsp::Url
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub ast: String
        }
    
        pub const METHOD: &'static str = "witcherscript-ide/debug/scriptAst";
    }

    /// Returns script file's CST (Concrete Syntax Tree) representation
    pub mod script_cst {
        use super::*;
    
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub script_uri: lsp::Url
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub cst: String
        }
    
        pub const METHOD: &'static str = "witcherscript-ide/debug/scriptCst";
    }

    /// Returns the content graph visualization in graphviz .dot format
    pub mod content_graph_dot {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub dot_graph: String
        }
    
        pub const METHOD: &'static str = "witcherscript-ide/debug/contentGraphDot";
    }

    /// Returns a human-readable report containing all code symbols in a given script file
    pub mod script_symbols {
        use super::*;

        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Parameters {
            pub script_uri: lsp::Url
        }
        
        #[derive(Debug, Serialize, Deserialize)]
        #[serde(rename_all = "camelCase")]
        pub struct Response {
            pub symbols: String
        }
    
        pub const METHOD: &'static str = "witcherscript-ide/debug/scriptSymbols";
    }
}
